lib/internal/process/promises.js

```diff
@@ -29,6 +40,10 @@ function hasRejectionToWarn() {
 }
 
 function promiseRejectHandler(type, promise, reason) {
+  if (state === undefined) {
+    const { getOptionValue } = require('internal/options');
+    state = states[getOptionValue('--unhandled-rejections')];
```

Comment:
This line gets executed on each promise rejection in case if the option is not set (see above for explanation).
Is that intended?

If yes -- why? Also, if yes -- it should better be explicit, imo, e.g.:
```js
if (state === undefined || state === states.unset) {
```

That would improve code clarity.

If no -- perhaps this needs to be cached instead?

Reply:
Note: not "each unhandled promise rejection", but "each promise rejection".

```console
MacBook-ChALkeR:node chalker$ ./node
> Promise.reject(32).catch(() => {})
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
Promise { <pending> }
> Promise.reject(32).catch(() => {})
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
Promise { <pending> }
> Promise.reject(32).then(()=>{}).then(()=>{}).catch(() => {})
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
Promise { <pending> }
> Promise.reject(32).then(()=>{}).then(()=>{}).catch(() => {})
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
Promise { <pending> }
> Promise.all(Array(10).fill().map(() => Promise.reject(42))).catch(() => {})
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
Promise { <pending> }
> getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
getOptionValue: --unhandled-rejections
```

Reply:
Good catch. This should indeed be set to a default instead of executing each time.

I personally would love to move this out in general but this seems the only `safe` place to call receive the flags value. @joyeecheung would it be possible to set up this code after the `pre-execution` phase so it's possible to access the flags directly? Not really, right?

Reply:
Why not cache the result of `getOptionValue('--unhandled-rejections')` on the first call?

Although `getOptionValue` is not really super expensive, it's just a get from a map and then a named property get from a object that's likely to be fast. If there are unhandled rejections I doubt performance impact like this is the concern.

Reply:
It should be fixed now.

Reply:
Upd: happens on each potentially unhandled promise rejection.

Reply:
@joyeecheung 
>  If there are unhandled rejections I doubt performance impact like this is the concern.

This method is called not only on unhandled rejections, but on rejections which could potentially be unhandled or rejections that were handled later.
E.g. `Promise.reject(40).catch(() => {})` calls this method _twice_.

Still it probably was not a significant performance hit in real-world apps.

But either way, that's fixed already =).